cns/restserver/util.go

```diff
@@ -350,11 +359,145 @@ func (service *HTTPRestService) removeToBeDeletedIPStateUntransacted(
 	return 0, ""
 }
 
+func (service *HTTPRestService) getNetworkContainersResponse(
+	req cns.GetNetworkContainerRequest,
+) (getNetworkContainersResponse []cns.GetNetworkContainerResponse) {
+	var (
+		networkContainerID          string
+		getNetworkContainerResponse cns.GetNetworkContainerResponse
+		exists                      bool
+		waitingForUpdate            bool
+	)
+
+	service.Lock()
+	defer service.Unlock()
+
+	switch service.state.OrchestratorType {
+	case cns.AzureFirstParty:
+		podInfo, err := cns.UnmarshalPodInfo(req.OrchestratorContext)
+		if err != nil {
+			getNetworkContainerResponse.Response.ReturnCode = types.UnexpectedError
+			getNetworkContainerResponse.Response.Message = fmt.Sprintf("Unmarshalling orchestrator context failed with error %v", err)
+			getNetworkContainersResponse = append(getNetworkContainersResponse, getNetworkContainerResponse)
+			return getNetworkContainersResponse
+		}
+
+		logger.Printf("pod info %+v", podInfo)
+
+		exists = false
+		ncidBytes, err := json.Marshal(service.state.ContainerIDByOrchestratorContext[podInfo.Name()+podInfo.Namespace()])
+		if err != nil {
+			logger.Errorf("Failed to marlshal NC set")
+			return nil
+		}
+
+		networkContainerID = string(ncidBytes)
+		if networkContainerID != "" {
+			exists = true
+		}
+
+		ncSet := &ncSet{}
+		if err = json.Unmarshal(ncidBytes, &ncSet); err != nil {
+			logger.Errorf("Failed to unmarlshal NC set")
+			return nil
+		}
+
+		if exists {
+			for ncid := range *ncSet {
+				// If the goal state is available with CNS, check if the NC is pending VFP programming
+				waitingForUpdate, getNetworkContainerResponse.Response.ReturnCode, getNetworkContainerResponse.Response.Message = service.isNCWaitingForUpdate(service.state.ContainerStatus[networkContainerID].CreateNetworkContainerRequest.Version, networkContainerID) //nolint:lll // bad code
+				// If the return code is not success, return the error to the caller
+				if getNetworkContainerResponse.Response.ReturnCode == types.NetworkContainerVfpProgramPending {
+					logger.Errorf("[Azure-CNS] isNCWaitingForUpdate failed for NCID: %s with error: %s",
+						ncid, getNetworkContainerResponse.Response.Message)
+					getNetworkContainersResponse = append(getNetworkContainersResponse, getNetworkContainerResponse)
+					return getNetworkContainersResponse
+				}
+
+				vfpUpdateComplete := !waitingForUpdate
+				ncstatus := service.state.ContainerStatus[ncid]
+				// Update the container status if-
+				// 1. VfpUpdateCompleted successfully
+				// 2. VfpUpdateComplete changed to false
+				if (getNetworkContainerResponse.Response.ReturnCode == types.NetworkContainerVfpProgramComplete &&
+					vfpUpdateComplete && ncstatus.VfpUpdateComplete != vfpUpdateComplete) ||
+					(!vfpUpdateComplete && ncstatus.VfpUpdateComplete != vfpUpdateComplete) {
```

Comment:
That these comments are needed hints at the fact that there's too much going on in this conditional. Consider moving this to maybe a function or some other type where you can break it up and document the individual parts.

Reply:
Need to redesign this part

Reply:
We can revisit this to optimize. But we can move this to a common function